File: java/src/main/java/com/myriad/wdl/model/definitions/WdlStruct.java
package com.myriad.wdl.model.definitions;

import com.myriad.wdl.model.WdlDocument.WdlDocumentElement;
import com.myriad.wdl.model.base.WdlNode;
import com.myriad.wdl.model.types.WdlType;
import java.util.ArrayDeque;
import java.util.Objects;
import java.util.Optional;
import lombok.Getter;
import lombok.Setter;

/**
 * Struct definition node.
 *
 * <p>The WDL specification describes structs as user-defined composite types whose members are named
 * and typed. This class models the struct declaration itself, while {@link WdlStructMember}
 * models each declared member.
 */
public final class WdlStruct implements WdlDocumentElement {
  /** Marker for any node that can appear directly inside a struct definition. */
  public interface WdlStructElement extends WdlNode {}

  @Getter @Setter private String name;
  private final ArrayDeque<WdlStructElement> elements = new ArrayDeque<>();

  public WdlStruct() {}

  public WdlStruct(String name) {
    setName(name);
  }

  /** Returns the ordered struct elements, usually member declarations and metadata sections. */
  public ArrayDeque<WdlStructElement> elements() {
    return elements;
  }

  /** Returns whether a member with the supplied name exists. */
  public boolean hasMember(String memberName) {
    return member(memberName).isPresent();
  }

  /** Returns the declared member, if present. */
  public Optional<WdlStructMember> member(String memberName) {
    if (memberName == null || memberName.isBlank()) {
      return Optional.empty();
    }
    return elements().stream()
        .filter(WdlStructMember.class::isInstance)
        .map(WdlStructMember.class::cast)
        .filter(member -> Objects.equals(memberName, member.getName()))
        .findFirst();
  }

  /** Returns the declared member type, if present. */
  public Optional<WdlType> memberType(String memberName) {
    return member(memberName).map(WdlStructMember::getType);
  }

  @Override
  public String toString() {
    return getClass().getSimpleName();
  }

  public static final class WdlStructMember implements WdlStructElement {
    @Getter @Setter private WdlType type;
    @Getter @Setter private String name;

    public WdlStructMember() {}

    public WdlStructMember(WdlType type, String name) {
      setType(type);
      setName(name);
    }

    @Override
    public String toString() {
      return getClass().getSimpleName();
    }
  }
}

File: python/src/wdl_model/model/definitions/wdl_struct.py
"""Struct definition nodes for the Python WDL object model."""

from __future__ import annotations

from collections import deque
from dataclasses import dataclass, field

from wdl_model.model.base import WdlNode
from wdl_model.model.types import WdlType


class WdlStructElement(WdlNode):
    """Marker for any node that can appear directly inside a struct definition."""

    pass


@dataclass
class WdlStructMember(WdlStructElement):
    """Named, typed member of a WDL struct."""

    type: WdlType | None = None
    name: str | None = None


@dataclass
class WdlStruct:
    """Struct definition node.

    Structs are user-defined composite types whose members are named and typed.
    """

    name: str | None = None
    _elements: deque[WdlStructElement] = field(default_factory=deque)

    def elements(self) -> deque[WdlStructElement]:
        """Return the ordered struct elements, usually members and metadata sections."""
        return self._elements

    def hasMember(self, member_name: str | None) -> bool:
        """Return whether a member with the supplied name exists."""
        return self.member(member_name) is not None

    def member(self, member_name: str | None) -> WdlStructMember | None:
        """Return the declared member by name, if present."""
        if member_name is None or not member_name.strip():
            return None
        for element in self._elements:
            if isinstance(element, WdlStructMember) and element.name == member_name:
                return element
        return None

    def memberType(self, member_name: str | None) -> WdlType | None:
        """Return the declared member type by name, if present."""
        member = self.member(member_name)
        return member.type if member is not None else None
